### ISSUE TITLE
vscode update breaks HLS  in Windows and Linux

### ISSUE THREAD

**Issue by KeesBleijenberg** [haskell/vscode-haskell#606]
When I start vscode (version 1.66.2)  in Linux to load a stack project, vscode suddenly says: Project requires HLS but it isn't installed. When I click on "Install HLS" it wants to install GhcUp. Do I have to install ghcup (which does more or less the same as stack?). When I upgrade ghc (another resolver) in stack does the HLS know this upgrade?
The problem is the same on Windows. Is this a vscode + stack + HLS problem or is it just a vscode problem?

Stack is on the path and has version 2.7.5
In the project directory: stack ghc -- --version => version 8.8.4
The setting for the plugin Manage HLS is set to PATH.

**Comment by fendor** [comment]
Hi, thank you for the bug report!

You **don't** need to mix ghcup and stack, but if you want to automatically install HLS, you will need ghcup.

However, it is perfectly possible to use ghcup and your already installed stack version!
After you have successfully installed ghcup, you may add this:
```json
{
  "haskell.toolchain": {
    "ghc": null,
    "cabal": null,
    "stack": null
  }
}
```

to your global VSCode `settings.json` file. (See https://github.com/haskell/vscode-haskell#setting-a-specific-toolchain for more examples).

Then ghcup will only be used to install HLS and VSCode will use your already installed stack version.
And yes, the extension will still automatically install the correct version of HLS depending on your project!

**Comment by KeesBleijenberg** [comment]
Then I tried to install ghcUp. The installation on Linux and Windows both failed :-(

This is the log on linux:
Debug: Identified Platform as: Linux Mint, 19.3
Debug: last access was 19.572510329s ago, cache interval is 300s
Debug: Decoding yaml at: /home/kees/.ghcup/cache/ghcup-0.0.7.yaml
Debug: Requested to install GHC with 8.10.7
Info: downloading: https://downloads.haskell.org/~ghc/8.10.7/ghc-8.10.7-x86_64-deb10-linux.tar.xz as file /home/kees/.ghcup/cache/ghc-8.10.7-x86_64$
Info: verifying digest of: ghc-8.10.7-x86_64-deb10-linux.tar.xz
Info: Unpacking: ghc-8.10.7-x86_64-deb10-linux.tar.xz to /tmp/ghcup-3b53efb243b823e9
Info: Installing GHC (this may take a while)
Debug: Running sh with arguments ["./configure","--prefix=/home/kees/.ghcup/ghc/8.10.7"]
Debug: Running make with arguments ["install"]
Error: (Process "make" with arguments ["install"] failed with exit code 2.,
 ())
Error: Also check the logs in /home/kees/.ghcup/logs

On Windows, I have to run a Powershell script. Unfortunately  this script does not work on windows 7 (I know very old). The error:

PS C:\Users\Kees> Set-ExecutionPolicy Bypass -Scope Process -Force;[System.Net.S
ervicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::Securi
tyProtocol -bor 3072;Invoke-Command -ScriptBlock ([ScriptBlock]::Create((Invoke-
WebRequest https://www.haskell.org/ghcup/sh/bootstrap-haskell.ps1 -UseBasicParsi
ng))) -ArgumentList $true
Exception setting "SecurityProtocol": "Cannot convert value "3312" to type "Sys
tem.Net.SecurityProtocolType" due to invalid enumeration values. Specify one of
 the following enumeration values and try again. The possible enumeration value
s are "Ssl3, Tls"."
At line:1 char:84
+ Set-ExecutionPolicy Bypass -Scope Process -Force;[System.Net.ServicePointMana
ger]:: <<<< SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtoc
ol -bor 3072;Invoke-Command -ScriptBlock ([ScriptBlock]::Create((Invoke-WebRequ
est https://www.haskell.org/ghcup/sh/bootstrap-haskell.ps1 -UseBasicParsing)))
-ArgumentList $true
    + CategoryInfo          : InvalidOperation: (:) [], RuntimeException
    + FullyQualifiedErrorId : PropertyAssignmentException

The term 'Invoke-WebRequest' is not recognized as the name of a cmdlet, functio
n, script file, or operable program. Check the spelling of the name, or if a pa
th was included, verify that the path is correct and try again.
At line:1 char:233
+ Set-ExecutionPolicy Bypass -Scope Process -Force;[System.Net.ServicePointMana
ger]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bo
r 3072;Invoke-Command -ScriptBlock ([ScriptBlock]::Create((Invoke-WebRequest <<
<<  https://www.haskell.org/ghcup/sh/bootstrap-haskell.ps1 -UseBasicParsing)))
-ArgumentList $true
    + CategoryInfo          : ObjectNotFound: (Invoke-WebRequest:String) [], C
   ommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

**Comment by hasufell** [comment]
Windows 7 support isn't very good at the moment (due to lack of testers etc.): https://gitlab.haskell.org/haskell/ghcup-hs/-/issues/140

You can try using [chocolatey](https://community.chocolatey.org/packages?q=Haskell) instead and then tell the extension to use your system PATH:

```json
"haskell.manageHLS": "PATH"
```

Alternatively, install only [msys2](https://community.chocolatey.org/packages/msys2) via chocolatey, then download the [ghcup binary](https://downloads.haskell.org/~ghcup/x86_64-mingw64-ghcup.exe) manually into e.g. `C:\ghcup\bin\ghcup.exe` and then set the env var `GHCUP_MSYS2` to `C:\tools\mingw64` (wherever chocolatey installs msys2 to).

Then follow https://www.haskell.org/ghcup/guide/#basic-usage

**Comment by KeesBleijenberg** [comment]
I've attched the .ghcUp logs:
[ghc-configure.log](https://github.com/haskell/vscode-haskell/files/8599155/ghc-configure.log)
[ghc-make.log](https://github.com/haskell/vscode-haskell/files/8599158/ghc-make.log)
[ghcup.log](https://github.com/haskell/vscode-haskell/files/8599159/ghcup.log)

**Comment by hasufell** [comment]
> /home/kees/.ghcup/ghc/8.10.7/lib/ghc-8.10.7/bin/ghc-pkg: error while loading shared libraries: libtinfo.so.6: cannot open shared object file: No such file

It appears linux mint 19.3 is on an old ncurses, so the tarball ghcup picks is incorrect. Give me 30 minutes and I'll fix it.

**Comment by hasufell** [comment]
https://github.com/haskell/ghcup-metadata/commit/abf57dbe8ca756c39ce5c61a36646f3fb0c6a38f

**Comment by KeesBleijenberg** [comment]
The Linux problem is solved. Thanks!!

In the past I had a lot of problems installing mingw (or msys2). The network package, if recall correctly. I also had problems installing chocolatey.  
I still wonder what went wrong. VS code and HLS worked for a long time on Linux and Windows. Suddenly it breaks and I have to investigate HLS, ghcUp, chocolatey and even msys2. A mystery to me. 

Can I help testing (a new script?) for installing HLS on windows 7?

**Comment by hasufell** [comment]
> Can I help testing (a new script?) for installing HLS on windows 7?

The script is here: https://gitlab.haskell.org/haskell/ghcup-hs/-/blob/master/scripts/bootstrap/bootstrap-haskell.ps1

If you can make it work on windows 7, I'll update it.

**Comment by KeesBleijenberg** [comment]
I rarely use Powershell so maybe I'am doing stupid things.
First I tried to double click the script in explorer. => red text flashes and nothing happens.
Then I opend a cmd, started powershell and then on the prompt Set-ExecutionPolicy ByPass. Powershell complains that access to a registry key xxxx is denied. All alternatives for ByPass fail the same way.

Second attempt: open a cmd as administrator, start powershell and now Set-ExecutionPolcy ByPass succeeds. 
Then on the powshell prompt I do .\bootstrap-haskell.ps1 <enter>

This is the error I got:
PS D:\ghcUp> .\bootstrap-haskell.ps1
The term 'Get-CimInstance' is not recognized as the name of a cmdlet, function,
 script file, or operable program. Check the spelling of the name, or if a path
 was included, verify that the path is correct and try again.
At D:\ghcUp\bootstrap-haskell.ps1:204 char:32
+   $partitions = Get-CimInstance <<<<  win32_logicaldisk
    + CategoryInfo          : ObjectNotFound: (Get-CimInstance:String) [], Par
   entContainsErrorRecordException
    + FullyQualifiedErrorId : CommandNotFoundException

Looks like Get_CimInstance is broken on Window 7.

I'am using a 64 bit version of Windows.

**Comment by KeesBleijenberg** [comment]
Meanwhile I updated powershell (not so easy on win 7) to version 5.1 
The I ran the script again and got different error messages: 

PS D:\ghcUp> Set-ExecutionPolicy ByPass

Execution Policy Change
The execution policy helps protect you from scripts that you do not trust. Ch
you to the security risks described in the about_Execution_Policies help topi
http://go.microsoft.com/fwlink/?LinkID=135170. Do you want to change the exec
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (defaul
PS D:\ghcUp> .\bootstrap-haskell.ps1
Setting env variable GHCUP_INSTALL_BASE_PREFIX to 'C:\'
Preparing for GHCup installation...
GHCup already installed at 'C:\\ghcup'...
Continuing installation...
First checking for Msys2...
...Msys2 doesn't exist, installing into C:\\ghcup\msys64
Starting installation in 5 seconds, this may take a while...
Downloading Msys2 archive...
Get-FileWCSynchronous : Exception calling "DownloadFile" with "2" argument(s)
te SSL/TLS secure channel."
At D:\ghcUp\bootstrap-haskell.ps1:411 char:7
+       Get-FileWCSynchronous -url ('https://repo.msys2.org/distrib/{0} ...
+       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Get-FileWCSynchronous], Meth
    + FullyQualifiedErrorId : WebException,Get-FileWCSynchronous

**Comment by KeesBleijenberg** [comment]
Progress.
I put in the script the line: [System.Net.ServicePointManager]::SecurityProtocol = "Tls12"
When I run the script msys2 and ghcup is installed. In the terminal in  VScode I can do ghcup --version and get v0.1.17.7 
No questions about HLS are asked. 
I've added these lines in  settings.json 
   "haskell.manageHLS": "PATH",
   "http.proxyAuthorization": null
   "haskell.toolchain": {
    "ghc": null,
    "cabal": null,
    "stack": null
  }

Unfortunately vscode still complains HLS is not installed. After a click on install HLS the web page for ghcUp is opened as if ghcup is not installed. What can I do?

**Comment by hasufell** [comment]
@KeesBleijenberg 

You need to set `"haskell.manageHLS": "GHCup"`.

Also, please share the final result of your powershell script.

**Comment by KeesBleijenberg** [comment]
I inserted   "haskell.manageHLS": "GHCup" in settings.json and it works. After starting VScode and clicking on 'install HLS', ghcup installs the language server. 

About the script. To be honest, I know almost nothing about Powershell. I guessed that the problem might be that Powershell uses an old version of TLS. I found the line 
[System.Net.ServicePointManager]::SecurityProtocol = "Tls12"
for let Powershell use to a newer version of TLS.
I couldn't find a 'main' in the powershell script, so I did prepend the line to the script.
After restarting the script I got (as expected) syntax errors, but the script (unexpected) continued and started installing msys2 and ghcUp.
After installing I was looking for a good place for this line and tried: 
...
if ($msys2Decision -eq 0) {
   [System.Net.ServicePointManager]::SecurityProtocol = "Tls12"
    Print-Msg -msg ('...Msys2 doesn''t exist, installing into {0}' -f $MsysDir)

	Print-Msg -msg 'Starting installation in 5 seconds, this may take a while...'
	Start-Sleep -s 5
....
When I run the script, I don't get syntax errors and it runs without problems. I don't know whether it still works okay, because msys2 and ghcUp are already installed.
